yksgi/src/lib.rs

```diff
@@ -0,0 +1,344 @@
+                LLVMOpcode::LLVMRet => {
+                    if self.ret() {
+                        return true;
+                    }
+                },
+                LLVMOpcode::LLVMStore => self.store(),
+                LLVMOpcode::LLVMSwitch => {
+                    self.switch();
+                    continue;
+                }
+                _ => todo!("{:?}", self.pc.as_str()),
+            }
+            self.pc = LLVMValue(LLVMGetNextInstruction(self.pc.0));
+        }
+    }
+
+    unsafe fn add(&mut self) {
+        let op1 = self.pc.get_operand(0);
+        let op2 = self.pc.get_operand(1);
+        let val1 = self.lookup(&op1);
+        let val2 = self.lookup(&op2);
+        // FIXME: Can overflows happen? What do we do?
+        let res = val1.val + val2.val;
+        self.setvar(self.pc, SGValue::new(res, self.pc.get_type()));
+    }
+
+    /// Convert value to new type of same bit size (without changing any bits).
+    pub unsafe fn bitcast(&mut self) {
+        // Since we store all integer values as u64 and evaluate them lazily according
+        // to the attached type, this operation is a NOP.
+        let src = LLVMValue(LLVMGetOperand(self.pc.0, 0));
+        let dstty = LLVMType(LLVMTypeOf(self.pc.0));
+        let val = self.lookup(&src);
+        self.setvar(self.pc, val.with_type(dstty));
+    }
+
+    /// Interpret branch instruction `instr`.
+    pub unsafe fn branch(&mut self) {
+        debug_assert!(self.pc.is_br());
+        let succ = if LLVMIsConditional(self.pc.0) == 0 {
+            LLVMBasicBlock::new(LLVMGetSuccessor(self.pc.0, 0))
+        } else {
+            let cond = LLVMValue::new(LLVMGetCondition(self.pc.0));
+            let val = self.lookup(&cond);
+            if val.val == 1 {
+                LLVMBasicBlock::new(LLVMGetSuccessor(self.pc.0, 0))
+            } else {
+                LLVMBasicBlock::new(LLVMGetSuccessor(self.pc.0, 1))
+            }
+        };
+        self.lastbb = Some(LLVMBasicBlock::new(LLVMGetInstructionParent(self.pc.0)));
+        self.pc = succ.first();
+    }
+
+    pub unsafe fn call(&mut self) {
+        // FIXME: Properly implement calls.
+        let func = LLVMGetCalledValue(self.pc.0);
+        let mut size = 0;
+        let name = CStr::from_ptr(LLVMGetValueName2(func, &mut size)).to_str().unwrap();
+        if name.starts_with("llvm.lifetime") {
+            return;
+        } else if name.starts_with("puts") || name.starts_with("printf") {
+            let op = self.pc.get_operand(0);
+            let op2 = op.get_operand(0);
+            let op3 = LLVMGetInitializer(op2.0);
+            let mut l = 0 as usize;
+            let s = LLVMGetAsString(op3, &mut l);
+            //libc::write(2, s as *const libc::c_void, l as libc::size_t - 1);
+            libc::puts(s);
+        } else if name.starts_with("__ykrt_control_point"){
+            todo!()
+        }
+        else {
+            // FIXME: For now just skip all other calls.
```

Comment:
Can we `todo` this branch?

Reply:
I would much rather whitelist the functions we want to skip, even if there are a lot of them, so that we don't find ourselves skipping things unexpectedly in the future. I'd just stick a `print` in here, run the tests, capture all of the function names, then say `if function not in <big array> { panic!(...) }`.

Reply:
Fixed in https://github.com/ykjit/yk/commit/e8a388b9053bac0fa3a8977bd59ba6d75b960457